orchagent/bfdorch.cpp

```diff
@@ -0,0 +1,441 @@
+};
+
+const map<sai_bfd_session_state_t, string> session_state_loopup =
+{
+    {SAI_BFD_SESSION_STATE_ADMIN_DOWN,  "Admin_Down"},
+    {SAI_BFD_SESSION_STATE_DOWN,        "Down"},
+    {SAI_BFD_SESSION_STATE_INIT,        "Init"},
+    {SAI_BFD_SESSION_STATE_UP,          "Up"}
+};
+
+BfdOrch::BfdOrch(DBConnector *db, string tableName, TableConnector stateDbBfdSessionTable):
+    Orch(db, tableName),
+    m_stateBfdSessionTable(stateDbBfdSessionTable.first, stateDbBfdSessionTable.second)
+{
+    SWSS_LOG_ENTER();
+
+    DBConnector *notificationsDb = new DBConnector("ASIC_DB", 0);
+    m_bfdStateNotificationConsumer = new swss::NotificationConsumer(notificationsDb, "NOTIFICATIONS");
+    auto bfdStateNotificatier = new Notifier(m_bfdStateNotificationConsumer, this, "BFD_STATE_NOTIFICATIONS");
+    Orch::addExecutor(bfdStateNotificatier);
+}
+
+BfdOrch::~BfdOrch(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+void BfdOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string key =  kfvKey(t);
+        string op = kfvOp(t);
+        auto data = kfvFieldsValues(t);
+
+        if (op == SET_COMMAND)
+        {
+            if (!create_bfd_session(key, data))
+            {
+                it++;
+                continue;
+            }
+        }
+        else if (op == DEL_COMMAND)
+        {
+            if (!remove_bfd_session(key))
+            {
+                it++;
+                continue;
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+        }
+
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+void BfdOrch::doTask(NotificationConsumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    std::string op;
+    std::string data;
+    std::vector<swss::FieldValueTuple> values;
+
+    consumer.pop(op, data, values);
+
+    if (&consumer != m_bfdStateNotificationConsumer)
+    {
+        return;
+    }
+
+    if (op == "bfd_session_state_change")
+    {
+        uint32_t count;
+        sai_bfd_session_state_notification_t *bfdSessionState = nullptr;
+
+        sai_deserialize_bfd_session_state_ntf(data, count, &bfdSessionState);
+
+        for (uint32_t i = 0; i < count; i++)
+        {
+            sai_object_id_t id = bfdSessionState[i].bfd_session_id;
+            sai_bfd_session_state_t state = bfdSessionState[i].session_state;
+
+            SWSS_LOG_INFO("Get BFD session state change notification id:%" PRIx64 " state: %s", id, session_state_loopup.at(state).c_str());
+
+            if (state != bfd_session_lookup[id].state)
+            {
+                auto key = bfd_session_lookup[id].peer;
+                m_stateBfdSessionTable.hset(key, "state", session_state_loopup.at(state));
+
+                SWSS_LOG_NOTICE("Updated BFD session state for %s to %s", key.c_str(), session_state_loopup.at(state).c_str());
```

Comment:
can you please rephrase this? `BFD session state for %s changed from %s to %s`. This shall be a critical log message

Reply:
Updated the log message.